### PR TITLE
refactor:  Advantages of TypeScript(#39520)

### PR DESCRIPTION
**Pull request by ajayrose25**

### CODE REVIEW

File: extraTs/constants.ts [hunk 1]
```diff
@@ -0,0 +1,10 @@
+export const permissions: Ipermissions = {
```
Comment (VisaaliSuccessive) [inline extraTs/constants.ts]: the defining interface must be like 
IPermissions not Ipermission

File: extraTs/index.ts [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+import { hasPermission, validateUsers } from './utils';
+diamond(3);
+equilateral(3);
+
+console.log(hasPermission(permissions.getUsers, 'trainee', 'read'));
+console.log(hasPermission(permissions.getUsers, 'trainee', 'write'));
+console.log(hasPermission(permissions.getUsers, 'trainer', 'read'));
+console.log(hasPermission(permissions.getUsers, 'trainee', 'all'));
+
+const users: Iusers = [
```
Comment (VisaaliSuccessive) [inline extraTs/index.ts:14]: where is the import statement of the user interface!

File: extraTs/constants.ts (hunk 1 above)
Comment (VisaaliSuccessive) [inline extraTs/constants.ts]: where is the import statement of the permissions interface!

File: extraTs/interfaces.ts [hunk 3]
```diff
@@ -0,0 +1,14 @@
+interface Ipermissions {
+  getUsers: Igetusers;
```
Comment (VisaaliSuccessive) [inline extraTs/interfaces.ts]: define the types for the inner object also here itself

File: package.json [hunk 4]
```diff
@@ -23,6 +30,15 @@
...
     "@babel/node": "^7.8.3",
-    "@babel/preset-env": "^7.8.3"
+    "@babel/preset-env": "^7.8.3",
+    "tsc-watch": "^4.1.0",
+    "typescript": "^3.7.5"
+  },
+  "dependencies": {
+    "@types/node": "^13.5.0",
+    "commitizen": "^4.0.3",
+    "husky": "^4.2.1",
```
Comment (VisaaliSuccessive) [inline package.json:45]: where is the husky implementation

File: extraTs/constants.ts (hunk 1 above)
Comment (ajayrose25) [reply extraTs/constants.ts]: Done

File: extraTs/index.ts (hunk 2 above)
Comment (ajayrose25) [reply extraTs/index.ts:14]: Done

File: extraTs/constants.ts (hunk 1 above)
Comment (ajayrose25) [reply extraTs/constants.ts]: Done

File: extraTs/interfaces.ts (hunk 3 above)
Comment (ajayrose25) [reply extraTs/interfaces.ts]: Done

File: package.json (hunk 4 above)
Comment (ajayrose25) [reply package.json:45]: Done

File: extraTs/index.ts [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+import { hasPermission, validateUsers } from './utils';
+diamond(3);
+equilateral(3);
+
+console.log(hasPermission(permissions.getUsers, 'trainee', 'read'));
+console.log(hasPermission(permissions.getUsers, 'trainee', 'write'));
+console.log(hasPermission(permissions.getUsers, 'trainer', 'read'));
+console.log(hasPermission(permissions.getUsers, 'trainee', 'all'));
+
+import { Iusers } from './interfaces';
```
Comment (VisaaliSuccessive) [inline extraTs/index.ts]: keep all the import statements at the top only

Comment (ajayrose25) [reply extraTs/index.ts]: Done

**Pull request merged**